File: server/internal/model/customer.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// Customer 是客户档案（customers 表，03-DATABASE.md:44-67）。
//
// balance_cents / points 只是查询缓存，事实来源分别是
// balance_transactions 与 points_transactions；任何余额变化必须同时产生流水。
type Customer struct {
	ID              int64          `gorm:"primaryKey" json:"id"`
	Name            string         `gorm:"size:64;not null" json:"name"`
	Phone           string         `gorm:"size:32" json:"phone"`
	Gender          string         `gorm:"size:16" json:"gender"`
	Birthday        *time.Time     `gorm:"type:date" json:"birthday"`
	Avatar          string         `gorm:"size:255" json:"avatar"`
	Wechat          string         `gorm:"size:64" json:"wechat"`
	Source          string         `gorm:"size:64" json:"source"`
	FirstVisitAt    *time.Time     `json:"first_visit_at"`
	LastVisitAt     *time.Time     `json:"last_visit_at"`
	TotalSpentCents int64          `gorm:"not null;default:0" json:"total_spent_cents"`
	BalanceCents    int64          `gorm:"not null;default:0" json:"balance_cents"`
	Points          int64          `gorm:"not null;default:0" json:"points"`
	Remark          string         `gorm:"type:text" json:"remark"`
	CreatedAt       time.Time      `json:"created_at"`
	UpdatedAt       time.Time      `json:"updated_at"`
	DeletedAt       gorm.DeletedAt `json:"-"`
}

// Tag 是客户标签（tags 表，03-DATABASE.md:69-77）；软删除不影响历史关联。
type Tag struct {
	ID        int64          `gorm:"primaryKey" json:"id"`
	Name      string         `gorm:"size:64;not null" json:"name"`
	Color     string         `gorm:"size:32" json:"color"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-"`
}

// CustomerTagRelation 是客户-标签关联（customer_tag_relations 表，03-DATABASE.md:79-85）。
//
// 表本身只有 customer_id/tag_id 两列，唯一联合索引防止重复挂标签。
type CustomerTagRelation struct {
	CustomerID int64 `gorm:"not null" json:"customer_id"`
	TagID      int64 `gorm:"not null" json:"tag_id"`
}

File: web/src/api/customer.ts
import { del, get, post, put } from './http'
import type { PageData } from './http'
import type { Tag } from './tag'

/**
 * 客户 DTO，对齐 server/internal/controller/customer.go 的 CustomerView。
 * 金额一律整数分，时间一律 UTC（ISO 字符串），生日按 YYYY-MM-DD 输出。
 */
export interface Customer {
  id: number
  name: string
  phone: string
  gender: string
  birthday: string | null
  avatar: string
  wechat: string
  source: string
  first_visit_at: string | null
  last_visit_at: string | null
  total_spent_cents: number
  balance_cents: number
  points: number
  remark: string
  /** 仅客户详情接口填充；列表接口恒为 []（不逐行查标签） */
  tags: Tag[]
  created_at: string
  updated_at: string
}

/** 新增/编辑客户的可写字段（POST/PUT /customers 请求体，后端白名单） */
export interface CustomerPayload {
  name: string
  phone: string
  gender: string
  birthday: string | null
  wechat: string
  source: string
  remark: string
}

/** GET /customers 查询参数（04-API.md:85-95）；keyword 支持姓名/手机号/微信号 */
export interface CustomerListQuery {
  keyword?: string
  phone?: string
  tag_id?: number
  sort?: string
  page?: number
  page_size?: number
}

/** GET /customers：分页 + 筛选（查询/新增/编辑 both，删除仅 admin，04-API.md:72） */
export function listCustomers(query: CustomerListQuery = {}): Promise<PageData<Customer>> {
  return get<PageData<Customer>>('/customers', { ...query })
}

/** POST /customers：非空手机号重复时后端返回 409（06-BUSINESS-RULES.md 第 1 节） */
export function createCustomer(payload: CustomerPayload): Promise<Customer> {
  return post<Customer>('/customers', payload)
}

/** GET /customers/:id：详情包含标签（含已软删除的历史标签） */
export function getCustomer(id: number): Promise<Customer> {
  return get<Customer>(`/customers/${id}`)
}

/** PUT /customers/:id：按 id 更新原行（改手机号不新建客户） */
export function updateCustomer(id: number, payload: CustomerPayload): Promise<Customer> {
  return put<Customer>(`/customers/${id}`, payload)
}

/**
 * DELETE /customers/:id：软删除，历史消费/充值/流水保留（06-BUSINESS-RULES.md:11）。
 * 路由层 RBAC 仅 admin 可调用；前端按角色隐藏按钮只是 UI 简化（07-UI.md:30）。
 */
export function deleteCustomer(id: number): Promise<Record<string, never>> {
  return del<Record<string, never>>(`/customers/${id}`)
}
